---
title: "Microbial Reference Material Template Report of Analysis"
author: "Nathan D. Olson"
date: "July 14, 2015"
output:
  pdf_document:
    fig_caption: yes
    latex_engine: xelatex
    toc: yes
  html_document: default
  word_document:
    fig_caption: yes
bibliography: micro_rm_references.bib
---

```{r setup, echo=FALSE, warning = FALSE, message = FALSE}
library(png)
library(jpeg)
library(grid)
library(peprr)
library(dplyr)
library(knitr)
require(kfigr)
opts_chunk$set(message=FALSE, warning=FALSE, echo = FALSE)
```


````{r metadata, echo = FALSE, message = FALSE}
source("rm_metadata.R")
source("calc_results_values.R")
```

```{r loadDB, echo=FALSE, message=FALSE} 
peprDB <- dplyr::src_sqlite(db_path)
```

# Introduction
## General background
Increasingly, high stakes decisions impacting public health and safety are being made using microbial genomic sequencing data [@Tang2014]. For example, whole genome sequencing was recently used as part of an investigation of the European 2011 Escherichia coli O1O4:H4 sprout associated outbreak [@Grad2012]. As the stakes increase so does the required level of confidence in the measurement. This reference material as well as three other microbial genomic DNA RM8375-8378 were developed to help advance the measurement assurance of microbial genomic sequencing and DNA sequencing in general. The development of these reference material was supported through an Interagency agreement with the Food and Drug Administration's Center for Devices, for use in validating DNA sequencing platforms for clinical applications.

## Material Description
<!--
Add text specific to rm regarding strain selection and publically available reference genome sequence
-->
```{r rmSpecificStrinInfo, echo = FALSE,child='rm_strain_selection_info.Rmd' }
```


Loftstrand Labs Limited[^lofstrand] grew a large batch of `r rm_strain` provided by `r strain_source`  to produce ~ `r (rm_vial_number * 3)/1000` mg of total extracted DNA, divided equally into `r rm_vial_number` vials (See appendix for RM label and manufacturer provided report of analysis). Each unit of `r rm_number` is approximately `r rm_mass` ug of extracted genomic DNA in Tris buffer.  The material is stored in 0.5 mL screw cap microcentrifuge tubes (Item number 1405-9710 USA Scientific [^usa]).

<!--
Add text specific to rm regarding culturing conditions and DNA extraction.
-->
```{r rmSpecificCulture, echo = FALSE,child='rm_specific_culture_ext.Rmd' }
```
 
This RM is isolated DNA rather than live cells because cells can mutate with each cell division, and the genome sequence may not be stable over time for live cells. Extracting DNA from a large batch of cells helps ensure that all vials contain essentially the same DNA sequences. Even though the DNA in the cells is likely to mutate at low frequency during the growth, the resulting mutations will be in extremely small proportions of the overall cells unless they confer a selective advantage and occur early in the culture process. Potential high frequency mutations will be identified as part of the material characterization process. Even if any mutations confer a selective advantage, the DNA from the large batch of cells was mixed before aliquoting to minimize the variability in the proportions of mutations among vials.
 
## Intended Use
The intended use of `r rm_number` is to help assess performance of high-throughput DNA sequencing.  The primary intended use case for the RM is DNA sequencing method validation and quality control.  The genomic DNA is intended to be analyzed in the same way as any other sample a laboratory would analyze extracted DNA, such as through the use of a genome assembly or variant calling bioinformatic pipeline (`r figr(label = "process", prefix = TRUE, link = TRUE, type = "Figure")`).  Because the RM is extracted DNA, it does not assess pre-analytical steps such as DNA extraction. It does however, challenge sequencing library preparation, sequencing machines, base calling algorithms, and the subsequent bioinformatic analysis.  This RM is not intended to assess bioinformatics steps such as strain identificaiton, phylogenetic analysis, or genome annotation. It is important to recognize the genome sequence is provided as an “Information Value” rather than “Certified Value” because we were not yet fully confident that we accounted for all known biases.  However, we evaluated the material using a number of orthogonal methods to minimize the impact of systematic errors associated with individual platforms. 

```{r process, echo = FALSE, fig.width = 6, fig.align='center', fig.cap= "Generic measurement process for genome assembly and variant calling intended use cases.  For sequencing quality control the results from the sequencing step would be compared to the reference material genome sequence to determine base level sequencing accuracy.", anchor = "Figure"}
grid.raster(readPNG("measurement_process.png"))
```
  

## Genome Sequence as a reference value
Characterizing a microbial genome RM at each of the $`r rm_genome_size/1000000`\times 10^6$ positions poses a significant metrological challenge. The consensus base call at each genomic position is a “nominal property”, because it is not a numerical value.  A previous sequence-based SRM 2374 was certified primarily based on capillary (or “Sanger”) sequencing, a relatively mature sequencing technology that is not easily scalable to whole genomes.  The maturity of capillary sequencing as well as the small size of SRM 2374 (< 100,000 bases) allowed for manual curation of capillary sequencing traces across the entire DNA sequences, and the assignment of qualitative (non-probabilistic) statements about the uncertainty of each base call.

To characterize `r rm_number`, we relied on a variety of less mature “Next Generation Sequencing” (NGS) technologies. Unlike capillary sequencing, NGS technologies can sequence a whole genome at a reasonable cost.  NGS technologies measure many reads at each position, from which variant callers and consensus base callers calculate probabilistic estimates of uncertainty based on Bayesian statistics and assuming binomial sampling.  Unfortunately, these probabilistic estimates of uncertainty are rarely accurate due to various sources of bias and error in the sequencing and bioinformatics.  Therefore, to develop a characterized reference genome sequence we first generated a _de-novo_ assembly of the genome then performed a base level analysis of the genome assembly.  

The sources of bias and error associated with different steps in the sequence measurement process are indicated in the cause-effect diagram below (`r figr(label = "causeEffect", prefix = TRUE, link = TRUE, type = "Figure")`). Please see [@Olson2015] for a detailed discussion of these sources of error.

```{r causeEffect, echo = FALSE, fig.align='center', fig.show = 'asis', fig.cap= "Cause and effect diagram indicating the sources of bias and error for the individual components of the sequence measurement process.", anchor = "Figure"}
grid.raster(readPNG("MicroRMCauseEffect.png"))
```

In addition to the sources of error and bias associated with the measurement there is an unknown degree of biological variability within the batch.  The genomic DNA was homogeneized prior to aliquoting into the individual vials but the biologically variability within the population of cells from which the DNA was extracted is unknown.  The mutations that make up this variability include single nucleotide polymorphisms (SNPs), insertions and deletions (indels), as well as structural variants.  Low frequency SNPs and indels can be mistaken as sequencing errors and low frequency structural variants can lead to ambiguities in the genome assembly.  Our characterization methods did not attempt to identify low frequency biologically variability within the RM batch, however, when interpreting the reference material characterization procedure results the potential for these low frequency mutations should be considered.
    
Genomic purity is provided as an information value and has not been evaluated for sources of bias and error.  The genomic purity of the reference material was assessed using a taxonomic sequence classification algorithm, the accuracy of the method was evaluated using simulated sequence data (Olson et al. _in prep_).  However, sources of bias associated with the measurement, namely accuracy of the database and assignment algorithm, have not been investigated.

\pagebreak

# Characterization Process
## Sequence Characterization
### Sequencing Measurements
Eight vials from the `r rm_number` lot were randomly selected as representatives of the lot for sequencing at NIST. In addition to NIST measurements, the material was sequenced using the Pacific Biosciences [^pacbio] RSII, and optical mapping was by OpGen through their MapIt service. 

<!--
Add text specific to rm regarding sequencing measurements made at NIST.
-->
```{r rmSpecificSequencing, echo = FALSE,child='rm_specific_nist_sequencing.Rmd' }
```

### Sequence Analysis Methods
For the genome assembly we started with a _de novo_ genome assembly and validated the assembly using optical mapping and sequencing data generated using orthogonal measurement methods. Long read sequence data generated with the Pacific Biosciences [^pacbio] RSII generate sequence reads with a median length of around 8kb, which is long enough to spand the ribosomal operon repeat regions, often resulting in a complete genome assembly [@Koren2013]. While the long reads allow for the assembly of a complete genome the relatively high error rate ~11\%[^pb-error] (primarily insertions and deletions) for individual reads can lead to small variant errors. 

The assembly was validated using orthogonal methods; optical mapping and short read sequencing.  Optical mapping provides information regarding the overall structure of the genome based on the location of restriction sites within long fragments (average size > 200Mb).  Comparison of the fragment patterns between the optical mapping data and _in-silico_ generated fragment patterns from the genome assembly indicates potential misassemblies in the genome assembly [@Mendelowitz2014]. The Optical mapping data was generated by a commercial vendor OpGen[^opgen] and sources of bias and error have not been fully investigated. The due to the resolution limit of the optical mapping method it is unable to detect misassemblies smaller than 3 kb.  

Potential misassemblies smaller than 3 kb were identified using short read data generated using orthogonal sequencing technologies Illumina[^illumina] MiSeq, and Ion Torrent[^iontorrent] PGM and Pilon [@Walker2014], software developed to identify and correct misassemblies in microbial genomes.  The sources of bias associated with each of the orthogonal methods were not fully investigated and agreement between the methods supports the hypothesis that the assembly is correct. 

Disagreements between the genome assembly and the optical mapping or short read results were investigated for sources of bias to account for the discrepancy. When appropriate the genome assembly was corrected.  If we were unable to identify the source of the bias that caused the misassembly the genome sequence is annotated appropriately to indicate region where the assembly is ambiguous. It is important to note that agreement between orthogonal methods does not guarantee a correct result as they may all be susceptible to the same source(s) of bias. We intend to update the genome assembly as new data is collected and errors in the assembly are revealed.
    
Base level analysis was performed by evaluating the proportion of sequencing reads with bases supporting the genome assembly base call at individual genome positions, within the the lot (base purity) and among individual vials (base homogeneity). Short read sequencing data generated using the MiSeq and PGM were used for base level analysis. Base purity was defined for the two platform individually, using samtools mpileup [@Li2009]. Base homogeneity, was assessed using varscan somatic variant caller [@Koboldt2009], using the methods described in RM8398 report of analysis. 

The purity of the reference material in terms of presence of DNA from sources other than `r rm_strain` was assessed using the metagenomic taxonomic read classification algorithm PathoScope [@Hong2014].  This method uses an expectation maximization algorithm where the sequence data are first mapped to a database comprised on all sequence data in the Genbank nt database, then through an iterative process re-assigns ambiguously mapped reads to based on the proportion of reads mapped unambigously to individual taxa in the database. 

#### Method Reproducibility
The methods used to characterize the reference material were developed in a manner to facilitate reuse and ensure computational reproducibility, obtaining the same results when starting with the same set of measurement data.  To achive this goal the data, computer code, as well as the computing environment are all publically available.  The raw sequence data is archived at the Genbank Sequence Read Archive, ([www.ncbi.nlm.nih.gov/sra](http://www.ncbi.nlm.nih.gov/sra)), see `r figr(label = "seqTable", prefix = TRUE, link = TRUE, type = "Table")` in the summary of sequencing data results section for accession numbers.  The code used to characterize the material is available through GitHub as a python bioinformatics pipeline, [www.github/nate-d-olson/pepr](http://www.github/nate-d-olson/pepr), and an R package, [www.github/nate-d-olson/peprr](http://www.github/nate-d-olson/peprr).  The bioinformatics pipeline performs the initial sequence analysis and uses a Docker environment for dependencies,[www.docker.org](http://www.docker.com). A flowchart of thepipeline used to characterize the RM can be found in Appendix. The R package, includes functions for loading the output from the pipeline into an sqlite database and generating the summary figures and tables found within this report of analysis.  

\pagebreak

### Sequence Analysis Results
#### Summary of Sequencing Datasets

<!--
Add text specific to rm regarding general summary of sequencing datasets.
-->
```{r rmSpecificDataSummary, child='rm_specific_dataset_summary.Rmd' }
```

```{r seq_table, anchor = "Figure"}
kable(seq_summary_table(peprDB), 
      round = 0, row.names = FALSE, caption = "Summary of sequencing datasets")
```

\pagebreak

#### Genome Assembly
<!--
Add text specific to rm regarding general summary of genome assembly
-->
```{r rmSpecificAssembly, child='rm_specific_assembly_summary.Rmd'}
```

<!--
Add text specific to rm regarding general summary of optical mapping results
-->
```{r rmSpecificOpticalMap, echo = FALSE,child='rm_specific_optical_map_summary.Rmd' }
```

```{r opgenAssembly, fig.cap= "Assembly of optical mapping data.", anchor = "Figure"}
grid.raster(readPNG("opgen_assembly.png"))
```

<!--
Add text specific to rm regarding general summary of assembly validation results
-->
```{r rmSpecificAssemblyValidation, child='rm_specific_assembly_validation_summary.Rmd' }
```

<!--
```{r assemblyComp, echo = FALSE, message=FALSE, warning=FALSE, fig.align='center', fig.cap= "Comparison of optical map data to genome assembly."}
# __SPECIFIC TO RM__
grid.raster(readPNG("opgen_assembly_comparison.png"))
```
-->

<!--
```{r pilonTable, echo =FALSE, message=FALSE, warning=FALSE}
## need a better way to present this information
```
-->

\pagebreak

#### Base Level Analysis
##### Base Purity
<!--
Add text specific to rm regarding general summary of base level purity.
-->
```{r rmSpecificBaseLevelPurity, echo = FALSE,child='rm_specific_base_level_purity_summary.Rmd' }
```


##### Sequence homogeneity
<!--
Add text specific to rm regarding general summary of base level homogeneity.
-->
```{r rmSpecificHomogeneity, echo = FALSE,child='rm_specific_base_level_homogeneity_summary.Rmd' }
```

```{r homogeneityTable, echo =FALSE, anchor = "Table"}
kable(homogeneity_sig_table(peprDB, rename_cols = TRUE), 
      row.names = FALSE,
      caption = "Pairwise variant analysis results")
```

```{r homogeneityPointFig, fig.cap= "Among sample variation in purity values.", anchor = "Figure"}
#not needed to support statements in report
#homogeneity_point_line_figure(peprDB)
```

#### Genomic Purity
<!--
Add text specific to rm regarding genomic purity.
-->
```{r rmSpecificGenomicPurity, echo = FALSE, child='rm_specific_genomic_purity.Rmd' }
```


```{r contamCountsFig, fig.cap= "Proportion of reads from contaminant DNA.", anchor = "Figure"}
contam_counts_figure(peprDB, rm_genus)
```

```{r contamPointLineFig, fig.height = 6, fig.cap= "Breakdown of contaminants by organism.", anchor = "Figure"}
contam_point_line_figure(peprDB, rm_genus)
```

\pagebreak

## DNA Fragment Size Stability
`r rm_number` is stable under the recommended storage conditions, although there is evidence that DNA fragment size may change over time, as indicated by the accelerated aging study.  Samples were exposed to aging outside those recommended for normal use in a controlled experimental setting.  The results of this experiments indicate the degradation, or fragmentation, of the DNA may come after advanced aging. 

### Description of Stability Study
Vials of RM material were selected at random to assess stability of the DNA, in terms of the distribution of the molecular weight of the DNA.  For the stability study, a random number generator was used to determine the vials to measure.

An isochronous stability study was performed to ensure integrity of DNA fragment size for candidate `r rm_number`over a variety of storage.  Two time courses and temperature conditions (4°C, and 37°C) were used to mimic accelerated aging. The eight-week time course was started first, followed by a two-week time course.  The study was conducted such that all time courses ended on the same day.  A -20°C temperature control was used since this is the recommended storage condition of candidate `r rm_number`. 

All temperature-based conditions were monitored with a data logger to track temperature, humidity and dew point over the entirety of the time-course.  All samples were stored in opaque boxes to minimize exposure the light.  At the conclusion of the time-course samples were placed in a -20°C freezer until measurements were made.

<!--
Add text specific to rm regarding stability experiment
-->
```{r rmSpecificStabExpDesign, child='rm_specific_stability_experimental_design.Rmd' }
```

### Analysis of Stability Study
To assess the stability of the DNA fragment size distribtion the proportion of DNA in a PFGE gel lane for each sample was determined for two regions between ladder markers. The high region is the proportion of DNA between the 48.5kb and 197kb markers and the low region between 9.24kb and 48.5kb markers. Treatment proportions were compared to controls using a T-test. The gels were processed using the R package peprrDnaStability ([www.github/nate-d-olson/peprDnaStability](http://www.github/nate-d-olson/peprDnaStability)). Prior to processing the gels were first manually cropped to remove regions of the gel to the left and right of the outer lanes. The `runfindMarkers` starts up an interactive application to define the parameters used in processing the gels. Cropped and original gel images as well as the parameter set used to analyze the gels are available at __%%TODO%%__.

#### Stability Study Results
<!--
Add text specific to rm regarding stability results
-->
```{r rmSpecificStability, child='rm_specific_stability_results.Rmd' }
```

\pagebreak


# Appendix
```{r lofstrandROA, fig.cap= "Certificate of analysis provided by the reference material manufacturer.", anchor = "Figure"}
# __SPECIFIC TO RM__
grid.raster(readPNG("Lofstrand_certificate_of_analysis.png"))
```


<!-- 
```{r echo = FALSE,child='rm_specific_vial_pull.Rmd' }
## Information about rm random vial selection
```
--> 


```{r rmlabel, fig.width=3, fig.height=3, fig.cap= "RM tube label", anchor = "Figure"}
# __SPECIFIC TO RM__
grid.raster(readPNG("RMLabel.png"))
```

\pagebreak

# References

[^opgen]: OpGen Inc. [opgen.com](http://opgen.com) 708 Quince Orchard Road Gaithersburg, MD 20878 USA
[^usa]: USA Scientific, Inc. [usascientific.com](http://www.usascientific.com) PO Box 3565 Ocala, FL 34478 USA
[^illumina]: Illumina Inc., [illumina.com/](http://www.illumina.com/) 5200 Illumina Way San Diego, CA 92122 USA
[^iontorrent]: Life Technologies Corp., [iontorrent.com/](http://www.iontorrent.com/) 7000 Shoreline Court #201, South San Francisco, CA 94080 USA
[^pacbio]: Pacific Biosciences of California Inc. [pacificbiosciences.com/](http://www.pacificbiosciences.com/) 1380 Willow Rd. Menlo Park, CA 94025 USA
[^rast]: Rapid Annotaiton Using Subsystem Technology (RAST) server [rast.nmpdr.org](http://rast.nmpdr.org/rast.cgi)
[^lofstrand]: Lofstrand Labs Limited 7961 Cessna Avenue, Gaithersburg, MD 20879 [lofstrand.com/](http://www.lofstrand.com/)
[^pb-error]: Manufacturer stated error rate, in "Understanding Accuracy in SMRT Sequencing" [http://www.pacificbiosciences.com](http://www.pacificbiosciences.com/pdf/Perspective_UnderstandingAccuracySMRTSequencing.pdf)
